File: scripts/kopiera-omrade.py
#!/usr/bin/env python3
"""Kopierar ett helt område (ämne) till en annan profil i Flippas databas.

Samma form som prio-klassning.py: ALLA dataoperationer kapslas här, så en session
bara behöver köra detta skript (ett stabilt kommandomönster → en permission-regel,
ingen ad-hoc curl). Anonym auth precis som appen – ingen hemlighet behövs.

  subjects                      lista områden (id, namn, ägare, lektioner, kort)
  plan     <subjectId> <profil> TORRKÖRNING: visar exakt vad som skulle skapas
  copy     <subjectId> <profil> skapar kopian och verifierar den mot källan

Kopian är VERBATIM: exakt samma objekt som ligger på servern, med bara "owner"
ändrad. Ingen normalisering kan gå fel. Lektions- och kort-id:n återanvänds –
de behöver bara vara unika inom sitt eget ämne, och favoriter/SRS nycklas på
ORDET, inte på id.

Säkerhet: skriptet gör exakt EN skrivning – en POST som skapar ett NYTT ämne.
Det finns ingen kodväg som ändrar eller raderar något befintligt. Källan läses.
Efter skrivningen läses kopian tillbaka och jämförs fält för fält mot källan.
"""
import json
import sys
import urllib.request
import urllib.error
import hashlib
import os
import re

API_KEY = "AIzaSyAFFQFMBqspO71R1ykDU6VdTSaFY1P-6dA"
DB = "https://flashcards-484e9-default-rtdb.europe-west1.firebasedatabase.app"


def die(msg):
    print(f"FEL: {msg}", file=sys.stderr)
    sys.exit(1)


def http(method, url, payload=None):
    data = json.dumps(payload).encode() if payload is not None else None
    req = urllib.request.Request(url, data=data, method=method,
                                 headers={"Content-Type": "application/json"})
    with urllib.request.urlopen(req) as r:
        return json.loads(r.read() or "null")


_token = None
def token():
    global _token
    if not _token:
        r = http("POST", f"https://identitytoolkit.googleapis.com/v1/accounts:signUp?key={API_KEY}",
                 {"returnSecureToken": True})
        _token = r["idToken"]
    return _token


def get(path):
    return http("GET", f"{DB}/{path}.json?auth={token()}")


def profiler():
    """Läser giltiga profil-id ur USERS i app.js – så skriptet aldrig hamnar i otakt."""
    här = os.path.dirname(os.path.abspath(__file__))
    src = open(os.path.join(här, "..", "app.js"), encoding="utf-8").read()
    block = re.search(r"const USERS = \[(.*?)\n\];", src, re.S)
    if not block:
        die("hittade inte USERS i app.js")
    return re.findall(r'id:\s*"([^"]+)"', block.group(1))


def matt(s):
    """Antal lektioner, kort och en checksumma över allt innehåll."""
    lektioner = list((s.get("lessons") or {}).items())
    kort = 0
    h = hashlib.sha256()
    for lid, l in sorted(lektioner):
        h.update(f"L:{lid}:{l.get('name')}\n".encode())
        for cid, c in sorted((l.get("cards") or {}).items()):
            kort += 1
            h.update(f"C:{cid}:{c.get('front')}:{c.get('back')}:"
                     f"{c.get('hint')}:{c.get('prio')}\n".encode())
    return len(lektioner), kort, h.hexdigest()[:16]


def cmd_subjects():
    alla = get("content/subjects") or {}
    print(f"\nOmråden ({len(alla)}):\n")
    for sid, s in sorted(alla.items(), key=lambda kv: kv[1].get("order", 0)):
        lek, kort, _ = matt(s)
        print(f"  {str(s.get('name') or '(namnlöst)'):<30} ägare: {str(s.get('owner') or '-'):<8}"
              f" {lek:>3} lektioner {kort:>5} kort   {sid}")
    print("\nSedan:  plan <subjectId> <profil>   (torrkörning)\n")


def _forbered(sid, agare):
    giltiga = profiler()
    if agare not in giltiga:
        die(f'okänd profil "{agare}". Giltiga enligt app.js: {", ".join(giltiga)}')
    alla = get("content/subjects") or {}
    if sid not in alla:
        die(f'inget område med id "{sid}" – kör "subjects" först')
    kalla = alla[sid]
    namn = (kalla.get("name") or "").strip().lower()
    for annat_id, s in alla.items():
        if s.get("owner") == agare and (s.get("name") or "").strip().lower() == namn:
            die(f'"{kalla.get("name")}" finns redan hos {agare} ({annat_id}) – ingen dubblett skapas')
    return kalla


def cmd_plan(sid, agare):
    kalla = _forbered(sid, agare)
    lek, kort, sig = matt(kalla)
    print(f"\nKälla : {kalla.get('name')}  (ägare {kalla.get('owner')}, id {sid})")
    print(f"Kopia : {kalla.get('name')}  (ägare {agare}, nytt id genereras)")
    print(f"Omfång: {lek} lektioner, {kort} kort · checksumma {sig}")
    print(f"Språk : {kalla.get('lang') or '(inget)'}")
    forsta = next(iter(sorted((kalla.get("lessons") or {}).items())), None)
    if forsta:
        prov = list(sorted((forsta[1].get("cards") or {}).items()))[:3]
        print(f'Stickprov ur "{forsta[1].get("name")}": '
              + " · ".join(f"{c.get('front')} = {c.get('back')}" for _, c in prov))
    print("\nTORRKÖRNING – inget skrevs. Kör 'copy' med samma argument för att göra det.\n")


def cmd_copy(sid, agare):
    kalla = _forbered(sid, agare)
    lek, kort, sig = matt(kalla)
    kopia = json.loads(json.dumps(kalla))   # verbatim djupkopia
    kopia["owner"] = agare

    svar = http("POST", f"{DB}/content/subjects.json?auth={token()}", kopia)
    nytt = (svar or {}).get("name")
    if not nytt:
        die("Firebase returnerade inget id – kontrollera i konsolen om något skrevs")

    # Verifiera: läs tillbaka kopian OCH källan, jämför
    ny = get(f"content/subjects/{nytt}") or {}
    kalla_nu = get(f"content/subjects/{sid}") or {}
    lek2, kort2, sig2 = matt(ny)
    lek3, kort3, sig3 = matt(kalla_nu)

    print(f"\nSKRIVET: {nytt}")
    print(f"  kopia : ägare {ny.get('owner')} · {lek2} lektioner · {kort2} kort · {sig2}")
    print(f"  källa : ägare {kalla_nu.get('owner')} · {lek3} lektioner · {kort3} kort · {sig3}")

    fel = []
    if ny.get("owner") != agare: fel.append("kopians ägare blev fel")
    if (lek2, kort2, sig2) != (lek, kort, sig): fel.append("kopian skiljer sig från källan")
    if (lek3, kort3, sig3) != (lek, kort, sig): fel.append("KÄLLAN ändrades – undersök omedelbart")
    if kalla_nu.get("owner") != kalla.get("owner"): fel.append("källans ägare ändrades")
    if fel:
        die(" · ".join(fel))
    print("OK: kopian är identisk med källan (samma checksumma) och källan är orörd.\n")


if __name__ == "__main__":
    args = sys.argv[1:]
    try:
        if not args: die(__doc__)
        cmd, rest = args[0], args[1:]
        if cmd == "subjects" and not rest: cmd_subjects()
        elif cmd == "plan" and len(rest) == 2: cmd_plan(*rest)
        elif cmd == "copy" and len(rest) == 2: cmd_copy(*rest)
        else: die("okänt kommando eller fel antal argument – kör utan argument för hjälp")
    except urllib.error.HTTPError as e:
        die(f"HTTP {e.code} mot Firebase: {e.read().decode()[:200]}")

File: scripts/kopiera-omrade.js
// Kopierar ett helt område (ämne) till en annan profil i Realtime Database.
// Samma mönster som send-push.js: REST + legacy DB-secret via miljövariabel.
//
//   export FIREBASE_DB_SECRET=...            (samma hemlighet som i GitHub Actions)
//   node scripts/kopiera-omrade.js                        # lista alla områden
//   node scripts/kopiera-omrade.js franska maria          # TORRKÖRNING (skriver inget)
//   node scripts/kopiera-omrade.js franska maria --skarp  # skriv på riktigt
//
// Kopian är VERBATIM: exakt samma objekt som ligger på servern, med bara "owner"
// ändrad. Ingen normalisering, inga nya id:n inuti – lektions- och kort-id:n är
// unika inom sitt eget ämne, och favoriter/SRS nycklas på ORDET, inte på id.
// Originalet läses bara, aldrig skrivs.

const DB = "https://flashcards-484e9-default-rtdb.europe-west1.firebasedatabase.app";
const SECRET = process.env.FIREBASE_DB_SECRET;

if (!SECRET) {
  console.error("Saknar FIREBASE_DB_SECRET. Sätt den i miljön först:");
  console.error("  export FIREBASE_DB_SECRET=...");
  process.exit(1);
}

const [fragment, agare, ...flaggor] = process.argv.slice(2);
const SKARP = flaggor.includes("--skarp");

// Profil-id:n måste matcha USERS i app.js – annars ser ingen området.
const PROFILER = ["tom", "hedvig", "wille", "karin", "martin", "maria", "guest"];

const rakna = (s) => {
  const lektioner = Object.values(s.lessons || {});
  return {
    lektioner: lektioner.length,
    kort: lektioner.reduce((n, l) => n + Object.keys(l.cards || {}).length, 0),
  };
};

async function las(sokvag) {
  const res = await fetch(`${DB}/${sokvag}.json?auth=${SECRET}`);
  if (!res.ok) throw new Error(`DB-läsfel ${res.status} på ${sokvag}`);
  return res.json();
}

async function main() {
  const alla = (await las("content/subjects")) || {};
  const poster = Object.entries(alla);

  if (!fragment) {
    console.log(`\nOmråden i databasen (${poster.length}):\n`);
    poster.forEach(([id, s]) => {
      const { lektioner, kort } = rakna(s);
      console.log(
        `  ${String(s.name || "(namnlöst)").padEnd(28)} ägare: ${String(s.owner || "-").padEnd(8)}` +
        ` ${String(lektioner).padStart(3)} lektioner ${String(kort).padStart(5)} kort   ${id}`
      );
    });
    console.log("\nKör igen med:  node scripts/kopiera-omrade.js <namnfragment> <profil>\n");
    return;
  }

  if (!agare) { console.error("Ange målprofil, t.ex.: ... franska maria"); process.exit(1); }
  if (!PROFILER.includes(agare)) {
    console.error(`Okänd profil "${agare}". Giltiga: ${PROFILER.join(", ")}`);
    console.error("Lägg till profilen i USERS i app.js först, annars ser ingen området.");
    process.exit(1);
  }

  const q = fragment.toLowerCase();
  const traffar = poster.filter(([, s]) => String(s.name || "").toLowerCase().includes(q));
  if (traffar.length === 0) { console.error(`Inget område matchar "${fragment}".`); process.exit(1); }
  if (traffar.length > 1) {
    console.error(`Flera områden matchar "${fragment}" – var mer specifik:`);
    traffar.forEach(([, s]) => console.error(`  - ${s.name} (ägare: ${s.owner || "-"})`));
    process.exit(1);
  }

  const [kallId, kalla] = traffar[0];
  const { lektioner, kort } = rakna(kalla);

  // Dubbelkopieringsspärr: har målprofilen redan ett område med samma namn?
  const redan = poster.find(([, s]) =>
    s.owner === agare && String(s.name || "").toLowerCase() === String(kalla.name || "").toLowerCase());
  if (redan) {
    console.error(`"${kalla.name}" finns redan hos ${agare} (${redan[0]}). Avbryter – ingen dubblett.`);
    process.exit(1);
  }

  console.log(`\nKälla:  ${kalla.name}  (ägare: ${kalla.owner || "-"}, id ${kallId})`);
  console.log(`Kopia:  ${kalla.name}  (ägare: ${agare}, nytt id genereras)`);
  console.log(`Omfång: ${lektioner} lektioner, ${kort} kort`);
  console.log(`Språk:  ${kalla.lang || "(inget)"}`);

  const kopia = JSON.parse(JSON.stringify(kalla)); // verbatim
  kopia.owner = agare;

  if (!SKARP) {
    console.log("\nTORRKÖRNING – inget skrevs. Lägg till --skarp för att göra det på riktigt.");
    const forsta = Object.values(kopia.lessons || {})[0];
    if (forsta) {
      const k = Object.values(forsta.cards || {}).slice(0, 3)
        .map((c) => `${c.front} = ${c.back}`).join(" · ");
      console.log(`Stickprov ur "${forsta.name}": ${k}`);
    }
    return;
  }

  const res = await fetch(`${DB}/content/subjects.json?auth=${SECRET}`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(kopia),
  });
  if (!res.ok) { console.error("Skrivfel", res.status, await res.text()); process.exit(1); }
  const nyttId = (await res.json()).name;

  // Läs tillbaka och kontrollera att kopian blev komplett
  const kontroll = await las(`content/subjects/${nyttId}`);
  const k2 = rakna(kontroll || {});
  const ok = kontroll && kontroll.owner === agare
    && k2.lektioner === lektioner && k2.kort === kort;
  console.log(`\nSkrivet: ${nyttId}`);
  console.log(`Kontroll: ägare ${kontroll && kontroll.owner}, ${k2.lektioner} lektioner, ${k2.kort} kort`);
  console.log(ok ? "✓ Kopian är komplett och originalet orört."
                 : "✗ Kopian stämmer INTE med källan – granska i konsolen innan du litar på den.");
  if (!ok) process.exit(1);
}

main().catch((e) => { console.error(e.message || e); process.exit(1); });
